### PR TITLE
PHP Config profiles

### PR DESCRIPTION
**Pull request by loic425**
Referenced issue #1539

Example:
```php
use Behat\Config\Config;
use Behat\Config\Profile;

$config = new Config();
$config->withProfile(new Profile('default', [
    'extensions' => [
        'some_extension' => [],
    ],
]));

return $config;
```

```php
use Behat\Config\Config;
use Behat\Config\Profile;

$config = new Config(['imports' => ['pretty.yml']]);
$config
    ->withProfile(new Profile('default', [
        'formatters' => [
            'pretty' => false,
            'progress' => null,
        ],
    ]))
    ->withProfile(new Profile('pretty_without_paths', [
        'formatters' => [
            'progress' => false,
            'pretty' => [
                'paths' => false,
            ],
        ],
    ]))
;
return $config;
```

### CODE REVIEW

**Comment by loic425** [comment]
@carlos-granados I didn't add any Behat tests, do you think we'll need some? cause it will be used via withExtension and withSuite in the Behat tests later.

**Comment by carlos-granados** [comment]
> @carlos-granados I didn't add any Behat tests, do you think we'll need some? cause it will be used via withExtension and withSuite in the Behat tests later.

I think that, since this is going to be merged independently, it should have its own tests

**Comment by loic425** [comment]
> > @carlos-granados I didn't add any Behat tests, do you think we'll need some? cause it will be used via withExtension and withSuite in the Behat tests later.
> 
> I think that, since this is going to be merged independently, it should have its own tests

done :)

**Comment by loic425** [comment]
> LGTM, thanks @loic425

Thx, I've detected an issue, I will submit it on this PR.

File: src/Behat/Testwork/ServiceContainer/Configuration/ConfigurationLoader.php [hunk 1]
```diff
@@ -200,19 +200,25 @@ protected function loadFileConfiguration($configPath, $profile)
...
-            $config = $phpConfig->toArray();
+            $config = $this->getPHPConfigObject($configPath)->toArray();
         } else {
             $config = (array) Yaml::parse(file_get_contents($configPath));
         }
 
         return $this->loadConfigs($basePath, $config, $profile);
     }
 
+    private function getPHPConfigObject(string $configPath): ConfigInterface
```
Comment (loic425) [inline src/Behat/Testwork/ServiceContainer/Configuration/ConfigurationLoader.php:211]: @carlos-granados It prevents var collisions.
Now, the only possible collision is with "$configPath" in the throw exception.

Comment (carlos-granados) [reply src/Behat/Testwork/ServiceContainer/Configuration/ConfigurationLoader.php:211]: @loic425 I really don't understand why this change is needed, can you please explain? 🤔

File: features/extensions.feature [hunk 2]
```diff
@@ -4,13 +4,26 @@ Feature: Extensions
...
-        extensions:
-          custom_extension.php:
-            param1: val1
-            param2: val2
+      <?php
+
+      use Behat\Config\Config;
+      use Behat\Config\Profile;
+
+      $profile = new Profile('default', [
```
Comment (loic425) [inline features/extensions.feature:14]: I've added a `$profile` var here that proves the vars collision is fixed.

File: src/Behat/Testwork/ServiceContainer/Configuration/ConfigurationLoader.php (hunk 1 above)
Comment (loic425) [reply src/Behat/Testwork/ServiceContainer/Configuration/ConfigurationLoader.php:211]: yeah indeed...

When using the "require" method. We override existing arguments from the current method (here the $profile var)

Comment (carlos-granados) [reply src/Behat/Testwork/ServiceContainer/Configuration/ConfigurationLoader.php:211]: Ahh, I get it now. Thanks!

**Comment by loic425** [comment]
@acoulton Could you please make a review on this one too?
We're going on fast on the whole PHP config, that's great. Thanks a lot to both of you.

File: src/Behat/Testwork/ServiceContainer/Configuration/ConfigurationLoader.php (hunk 1 above)
Comment (stof) [reply src/Behat/Testwork/ServiceContainer/Configuration/ConfigurationLoader.php:211]: if you want to reduce conflict even more, you could make this private method a pure wrapper isolating `require` by returning `mixed`, and validating the ConfigInterface type in the caller.

Also, this should probably be a static method to avoid giving access to `$this` in the required file. Or maybe even stricter by using a bound closure to also remove access to `self` (and so also potential access to private state of the current class): https://github.com/composer/composer/blob/5eeba719d3f1388f68773c28e3971a7f441ee919/src/Composer/Autoload/ClassLoader.php#L575-L577

Comment (loic425) [reply src/Behat/Testwork/ServiceContainer/Configuration/ConfigurationLoader.php]: @stof I've made the change, I hope this is exactly why you have described it.

File: src/Behat/Config/ProfileInterface.php [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Behat\Config;
+
+interface ProfileInterface extends ConfigInterface
```
Comment (acoulton) [inline src/Behat/Config/ProfileInterface.php:7]: I think we should call all of these `XXXConfigInterface` (like you have `SuiteConfigInterface` in #1540). Otherwise for things like Suite, Extension, etc there is potential for confusion - at least if just glancing at the code - about whether you're expected to pass an actual instance of the extension rather than just a config object.

Equally, I am wondering if we definitely need separate interfaces and final classes for all of these, if the interface  for most of them is only going to have `name()` and `toArray()`. If the sugar methods are only defined / implemented on the final class then it will be harder for people to find those, and the interfaces add extra noise.

Since you have to use the classes we provide to get the sugar methods, and you always have the option to pass an array (or a class of your own that doesn't satisfy any interface) into the `$settings` of the container object, there are still plenty of ways you can customise the config without needing to be able to implement an interface?

The exception to this might be for extensions and formatters, where by definition we're not in control of the available config options so interfaces would be useful to allow extensions to define a custom config object of their own.

File: src/Behat/Config/Config.php [hunk 4]
```diff
@@ -11,6 +11,13 @@ public function __construct(
     ) {
     }
 
+    public function withProfile(ProfileInterface $profile): self
+    {
+        $this->settings[$profile->name()] = $profile->toArray();
```
Comment (acoulton) [inline src/Behat/Config/Config.php:16]: What if they have already defined a profile with that name? Should we perhaps throw rather than silently overwriting (bearing in mind we're in the context of a single config here, obviously imports are merged together later).

The examples in the feature files already suggest that the PHP config files might become a lot more verbose / have more variables etc than the YAML ones - I think that increases the risk of accidentally having a duplicate key, and I wonder if we should protect against that?

**Review by acoulton — CHANGES_REQUESTED** [review]
I'm finding it a bit hard to picture what the end-state of this will look like - based on the PRs open so far I have a slight concern that expressing everything with chained sugar methods (as opposed to named constructor arguments, for example) will make the config files become a lot more verbose and hard to follow.

Although I appreciate that sugar methods have the major advantage that it's very clear whether you've provided a value, which can't really be achieved even with optional method arguments.

I realise that routers, DI containers etc follow the sugar pattern, but the key difference is that they are part of the tested code, so a confusion over config should show up in CI. Whereas if the config for the test framework is not applied as the user expects, they could end up with CI giving false positives without realising.

I wonder if it's worth updating #1539 with a full worked example of a config with multiple profiles, suites, formatters, extensions etc so we can check we're happy with the overall shape of the API? 

It may well be this is the right path, but I'm worried that things might look OK when dealing with the initial simple objects and we only realise later that we've backed ourselves into a corner.

Particularly because we will want to keep releasing from master regularly, which means very soon we'll have to guarantee BC for any of the incremental PRs that get merged.

File: src/Behat/Config/ProfileInterface.php (hunk 3 above)
Comment (loic425) [reply src/Behat/Config/ProfileInterface.php:7]: @acoulton For the `XXXConfigInterface` part, I'm mitigated on that. This was an error on #1540.
If we use this suffix for the interface, we need a Config suffix on implementations SuiteConfig, ProfileConfig...

But maybe that's better... To be honest, naming is never easy :+1:

Comment (loic425) [reply src/Behat/Config/ProfileInterface.php:7]: For the final classes, they contains default helpers/sugar methods, and users can define its own ones. But maybe that's too much control.

File: src/Behat/Config/Config.php (hunk 4 above)
Comment (loic425) [reply src/Behat/Config/Config.php:30]: We could add an optional arguments `bool $overwrite = false`
But cause you are able to redefine the same profile configuration via another file, it will not work.

File: src/Behat/Testwork/ServiceContainer/Configuration/ConfigurationLoader.php (hunk 1 above)
Comment (loic425) [reply src/Behat/Testwork/ServiceContainer/Configuration/ConfigurationLoader.php]: @carlos-granados Cause the PR is going to be harder than expected, maybe should I open another PR for this specific fix?

File: src/Behat/Config/Config.php (hunk 4 above)
Comment (stof) [reply src/Behat/Config/Config.php:30]: No you cannot. Defining the same key twice in a Yaml config file is a Yaml syntax error.

Comment (loic425) [reply src/Behat/Config/Config.php:30]: I mean, in multiple yaml configuration files,  you are able to redefine a suite with same key as defined in another file.

File: src/Behat/Config/ProfileInterface.php (hunk 3 above)
Comment (acoulton) [reply src/Behat/Config/ProfileInterface.php:7]: > If we use this suffix for the interface, we need a Config suffix on implementations SuiteConfig, ProfileConfig...

I don't think we automatically *have to*. IMO they have different usecases:

* The interface is for when you're looking at the method you're calling to see what it accepts. Here, the priority is to make very clear to the end-user that `withExtension` takes an object with extension configuration, not e.g. an instance of an extension (which you might assume, if new).
* The concrete class is generally for when you're looking at / working in a config file. Here, it is already clear that you have config objects, and probably the priority is to keep the naming as concise as possible so that users are mostly looking at the actual config, rather than the boilerplate PHP structure.

So if we have interfaces, personally I would have `SuiteConfigInterface`, `ProfileConfigInterface`, `ExtensionConfigInterface` and then `SuiteConfig`, `ProfileConfig`, `ExtensionConfig` - or even just `Suite`, `Profile`, `Extension` (leaving the Config in the namespace).

Comment (acoulton) [reply src/Behat/Config/ProfileInterface.php:7]: > For the final classes, they contains default helpers/sugar methods, and users can define its own ones. But maybe that's too much control.

I think it adds complexity (and a lot of very similar interfaces) for not much benefit. These objects (especially the bits that are defined by the interface) are just DTOs, and I wouldn't usually implement an interface for that.

In reality, if users want to use any of our helper methods, they'll mostly have to use our DTOs, because it's only the DTOs that have the helpers. If they don't, they can just do something like:

```
return new Behat\Config\Config([
  'default' => (new MyCustomProfileConfigObject(...))->toArray()
])
```

On the other hand, for the more common case of a user wanting to use the default helpers, it would be much easier to see that `->withProfile` takes a `Profile` or `ProfileConfig` object, and click directly through to that in their IDE rather than clicking first to an interface and then see what implements it.

> But if we define interfaces for extensions, I think it would be better to have the same control on Profiles, Suites etc.

Extensions and formatters are special cases, because unlike Profiles, Suites, etc we have no knowledge or control of what config schema they take. So it would be very common that an extension author would want to provide a config object that you could pass to `->withExtension()`.

**Review by acoulton — APPROVED** [review]
Thanks for all the work on this @loic425, LGTM. I'll ask @carlos-granados to re-review as it's changed a bit since he looked.

File: src/Behat/Config/ProfileInterface.php (hunk 3 above)
Comment (loic425) [reply src/Behat/Config/ProfileInterface.php]: I removed the interface.

**Review by carlos-granados — APPROVED** [review]
Goog job @loic425 🎉
